Client/Frontend/Settings/SearchCustomEngineViewController.swift

```diff
@@ -0,0 +1,727 @@
+                    self.navigationItem.rightBarButtonItem?.isEnabled = false
+                    self.spinnerView.stopAnimating()
+                    self.isAutoAddEnabled = false
+                case .loading:
+                    self.navigationItem.rightBarButtonItem = UIBarButtonItem(customView: self.spinnerView)
+                    self.spinnerView.startAnimating()
+            }
+        }
+    }
+    
+    private func handleError(error: Error) {
+        let alert: UIAlertController
+        
+        if let searchError = error as? SearchEngineError {
+            switch searchError {
+                case .duplicate:
+                    alert = ThirdPartySearchAlerts.duplicateCustomEngine()
+                case .invalidQuery:
+                    alert = ThirdPartySearchAlerts.incorrectCustomEngineForm()
+                case .failedToSave:
+                    alert = ThirdPartySearchAlerts.failedToAddThirdPartySearch()
+                case .missingInformation:
+                    alert = ThirdPartySearchAlerts.missingInfoToAddThirdPartySearch()
+            }
+        } else {
+            alert = ThirdPartySearchAlerts.failedToAddThirdPartySearch()
+        }
+        
+        log.error(error)
+        present(alert, animated: true, completion: nil)
+    }
+
+    // MARK: Actions
+    
+    @objc func checkAddEngineType(_ nav: UINavigationController?) {
+        if isAutoAddEnabled {
+            addOpenSearchEngine()
+        } else {
+            addCustomSearchEngine()
+        }
+    }
+    
+    func addCustomSearchEngine() {
+        view.endEditing(true)
+        
+        guard let title = titleText,
+              let urlQuery = urlText,
+              !title.isEmpty,
+              !urlQuery.isEmpty else {
+            present(ThirdPartySearchAlerts.missingInfoToAddThirdPartySearch(), animated: true, completion: nil)
+            return
+        }
+        
+        changeAddButton(for: .disabled)
+        addSearchEngine(with: urlQuery, title: title)
+    }
+    
+    @objc func cancel() {
+        navigationController?.popViewController(animated: true)
+    }
+}
+
+// MARK: - UITableViewDelegate UITableViewDataSource
+
+extension SearchCustomEngineViewController: UITableViewDelegate, UITableViewDataSource {
+    
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return Section.allCases.count
+    }
+    
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return 1
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        switch indexPath.section {
+            case Section.url.rawValue:
+                guard let cell =
+                        tableView.dequeueReusableCell(withIdentifier: Constants.urlInputRowIdentifier) as? URLInputTableViewCell else {
+                    return UITableViewCell()
```

Comment:
Add an `assertionFailure` here since returning `UITableViewCell` at runtime will crash regardless

Reply:
After adding `TableViewReusable`, the necessity for guarding is not required.